### PR TITLE
Faster hot reload

### PR DESCRIPTION
**Pull request by goderbauer**
This implements the `DartDependencySetBuilder` completely in Dart instead of calling out to `sky_snapshot` (Linux/Mac) or `gen_snapshot` (Windows) and allows us to use the same code path on all supported host platforms.

It also slightly reduces hot reload times on Linux from ~750ms to ~690ms for the unchanged flutter_gallery app and significantly reduces hot reload times on Windows from almost 1.5s to just slightly slower than on Linux.

This change will also allow us to retire `sky_snapshot` completely in the future.

### CODE REVIEW

File: packages/flutter_tools/test/data/dart_dependencies_test/syntax_error/foo.dart [hunk 1]
```diff
@@ -2,4 +2,4 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-bad programmer!
+import bad programmer!
```
Comment (Hixie) [inline packages/flutter_tools/test/data/dart_dependencies_test/syntax_error/foo.dart]: why this change?

**Comment by Hixie** [comment]
I am not familiar with the code, but this looks great.

**Comment by Hixie** [comment]
Does it make sense to add a test for the new code?

**Comment by Hixie** [comment]
LGTM if you need it, but you may wish to wait for a more informed reviewer.

File: packages/flutter_tools/test/data/dart_dependencies_test/syntax_error/foo.dart (hunk 1 above)
Comment (goderbauer) [inline packages/flutter_tools/test/data/dart_dependencies_test/syntax_error/foo.dart]: With this change, we only parse the directive section of dart sources to figure out the dependencies.  This syntax error was technically after the directive section and is therefore ignored by the new DartDependencySetBuilder. Introducing the import statement makes the syntax error part of the directive section.

Now you might ask: "But what about syntax errors after the directive section during hot reload?" After the changed sources are pushed to device they are parsed (again) by the Dart VM on device, which will provide a clear and actionable error message if it encounters any syntax errors. So, everything is good here as well.

**Comment by goderbauer** [comment]
There is no rush. I definitely want @johnmccutchan to take a look at this.

About tests: This is already pretty well covered by the tests in `dart_dependencies_test.dart` and `dependency_checker_test.dart`.

**Comment by goderbauer** [comment]
This fixes https://github.com/flutter/flutter/issues/8513.

File: packages/flutter_tools/lib/src/dart/dependencies.dart [hunk 2]
```diff
@@ -2,89 +2,49 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'dart:convert';
+import 'package:analyzer/analyzer.dart';
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: Do we really want to depend on the analyzer?

File: packages/flutter_tools/test/data/dart_dependencies_test/syntax_error/foo.dart (hunk 1 above)
Comment (johnmccutchan) [inline packages/flutter_tools/test/data/dart_dependencies_test/syntax_error/foo.dart]: The old sky_snapshot based system would detect many syntax errors beyond imports on the host and not send it to the device for parsing. I think it's fine trading faster hot reload times for more device side errors.

**Review by johnmccutchan — APPROVED** [review]
Nice! I've left some notes but I think the speedup makes it worthwhile!

File: packages/flutter_tools/lib/src/dart/dependencies.dart (hunk 2 above)
Comment (goderbauer) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: This package (`flutter_tools`) already has a dependency on the analyzer to support `flutter analyze`, so this is not bringing in a new dependency to the package.

Was bringing in a new dependency your main concern? Or is there something else you're concerned about with regard to using the analyzer for this particular purpose?

Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/dart/dependencies.dart]: @goderbauer I was mostly concerned about bloating the initial download size but if we already depend on it, no big deal.

File: packages/flutter_tools/test/data/dart_dependencies_test/syntax_error/foo.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter_tools/test/data/dart_dependencies_test/syntax_error/foo.dart]: Ah that makes sense. I was confused as to which test this is (we have a bevy of tests that verify handling of bad syntax, it turns out.)

**Comment by goderbauer** [comment]
First benchmark run on device lab seems to verify this improvement:

hot-reload-success

DevFS sync time droped from over 600ms to 510ms.

**Comment by Hixie** [comment]
That's within the margin of error. We'll need more data before we can say for sure. (Look at the values on the left of that chart, for example. They're around the same.)
